components/scream/src/share/util/scream_family_tracking.hpp

```diff
@@ -0,0 +1,170 @@
+#ifndef SCREAM_FAMILY_TRACKING_CLASS
+#define SCREAM_FAMILY_TRACKING_CLASS
+
+#include "ekat/std_meta//ekat_std_enable_shared_from_this.hpp"
+#include "ekat/ekat_assert.hpp"
+
+#include <list>
+#include <memory>
+#include <type_traits>
+
+namespace scream
+{
+
+/*
+ * Class that handles tracking of parent/children instances of the same class type
```

Comment:
Neat. Unless we want to introduce a "family" as a first-class concept, maybe we could call this `TypeTracking` or `TypeBasedTracking`?

Reply:
I struggled with the name. I used 'family' only cause the class stores "parent" and "children", so I kept the analogy rolling. But a better name might help.

What about `PointerGraphTracking` or something with the word "graph" or "tree" in it?

Reply:
Technically, the graph might not be a tree, since I do not check that my parent is not me, nor one of my children (or grandchildren or...). In scream, the use case is quite limited, so I didn't feel the need for this check (though if used with circular deps, there would def be bugs). If we moved this file to ekat (which might be a possibility, given that it has no scream-based dep), then I might add checks.

Reply:
If it should be a tree to avoid circular refs, it seems like this check might be useful. And it may have a home in EKAT--the curiously recurring template pattern is not too bad a tax to pay for this kind of tracking.

Reply:
Ok. I might do this in a second time, cause I don't want to delay this PR too much, if that's ok. I will work on it right after it gets merged though.